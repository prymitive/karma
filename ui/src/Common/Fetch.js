import merge from "lodash.merge";

import promiseRetry from "promise-retry";

const CommonOptions = {
  mode: "cors",
  credentials: "include",
  redirect: "follow",
};

const FetchRetryConfig = {
  retries: 9,
  minTimeout: 2000,
  maxTimeout: 5000,
};

const FetchGet = async (uri, options, beforeRetry) =>
  await promiseRetry(
    (retry, number) =>
      fetch(
        uri,
        merge(
          {},
          {
            method: "GET",
          },
          CommonOptions,
          {
            mode: number <= FetchRetryConfig.retries ? "cors" : "no-cors",
          },
          options
        )
      ).catch((err) => {
        beforeRetry && beforeRetry(number);
        return retry(err);
      }),
    FetchRetryConfig
  );

export { CommonOptions, FetchGet, FetchRetryConfig };
