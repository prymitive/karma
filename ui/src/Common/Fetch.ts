import merge from "lodash.merge";

import promiseRetry from "promise-retry";

const CommonOptions = {
  mode: "cors",
  credentials: "include",
  redirect: "follow",
};

const FetchRetryConfig = {
  retries: 9,
  minTimeout: 2000,
  maxTimeout: 5000,
};

type PreRetryCallback = (number: number) => void;

const FetchGet = async (
  uri: string,
  options: RequestInit,
  beforeRetry: PreRetryCallback
): Promise<Response> =>
  await promiseRetry(
    (retry, number) =>
      fetch(
        uri,
        merge(
          {},
          {
            method: "GET",
          },
          CommonOptions,
          {
            mode: number <= FetchRetryConfig.retries ? "cors" : "no-cors",
          },
          options
        )
      ).catch((err) => {
        beforeRetry && beforeRetry(number);
        return retry(err);
      }),
    FetchRetryConfig
  );

export { CommonOptions, FetchGet, FetchRetryConfig };
