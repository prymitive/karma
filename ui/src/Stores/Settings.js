import { action } from "mobx";
import { localStored } from "mobx-stored";

class SavedFilters {
  config = localStored(
    "savedFilters",
    {
      filters: [],
      present: false
    },
    {
      delay: 100
    }
  );

  save = action(newFilters => {
    this.config.filters = newFilters;
    this.config.present = true;
  });

  clear = action(() => {
    this.config.filters = [];
    this.config.present = false;
  });
}

class FetchConfig {
  constructor(refresh) {
    this.config = localStored(
      "fetchConfig",
      { interval: refresh },
      { delay: 100 }
    );

    this.setInterval = action(newInterval => {
      this.config.interval = newInterval;
    });
  }
}

class AlertGroupConfig {
  options = Object.freeze({
    expanded: { label: "Always expanded", value: "expanded" },
    collapsedOnMobile: {
      label: "Collapse on mobile",
      value: "collapsedOnMobile"
    },
    collapsed: { label: "Always collapsed", value: "collapsed" }
  });

  constructor(renderCount, collapseState, colorTitleBar) {
    this.config = localStored(
      "alertGroupConfig",
      {
        defaultRenderCount: renderCount,
        defaultCollapseState: collapseState,
        colorTitleBar: colorTitleBar
      },
      { delay: 100 }
    );
  }

  update = action(data => {
    for (const [key, val] of Object.entries(data)) {
      this.config[key] = val;
    }
  });
}

class SilenceFormConfig {
  config = localStored("silenceFormConfig", { author: "" }, { delay: 100 });

  saveAuthor = action(newAuthor => {
    this.config.author = newAuthor;
  });
}

class GridConfig {
  options = Object.freeze({
    default: { label: "Use defaults from karma config file", value: "default" },
    disabled: { label: "No sorting", value: "disabled" },
    startsAt: { label: "Sort by alert timestamp", value: "startsAt" },
    label: { label: "Sort by alert label", value: "label" }
  });
  constructor(groupWidth) {
    this.config = localStored(
      "alertGridConfig",
      {
        sortOrder: this.options.default.value,
        sortLabel: null,
        reverseSort: null,
        groupWidth: groupWidth
      },
      { delay: 100 }
    );
  }
}

class FilterBarConfig {
  constructor(autohide) {
    this.config = localStored(
      "filterBarConfig",
      {
        autohide: autohide
      },
      {
        delay: 100
      }
    );
  }
}

class Settings {
  constructor(defaults) {
    let defaultSettings;
    if (defaults === undefined) {
      defaultSettings = {
        Refresh: 30 * 1000 * 1000 * 1000,
        HideFiltersWhenIdle: true,
        ColorTitlebar: false,
        MinimalGroupWidth: 420,
        AlertsPerGroup: 5,
        CollapseGroups: "collapsedOnMobile"
      };
    } else {
      defaultSettings = defaults;
    }

    this.savedFilters = new SavedFilters();
    this.fetchConfig = new FetchConfig(
      defaultSettings.Refresh / 1000 / 1000 / 1000
    );
    this.alertGroupConfig = new AlertGroupConfig(
      defaultSettings.AlertsPerGroup,
      defaultSettings.CollapseGroups,
      defaultSettings.ColorTitlebar
    );
    this.gridConfig = new GridConfig(defaultSettings.MinimalGroupWidth);
    this.silenceFormConfig = new SilenceFormConfig();
    this.filterBarConfig = new FilterBarConfig(
      defaultSettings.HideFiltersWhenIdle
    );
  }
}

export { Settings };
