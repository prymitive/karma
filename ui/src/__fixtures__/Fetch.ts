import { MockAlert, MockAlertGroup, MockSilence } from "__fixtures__/Alerts";
import {
  APIAlertsResponseT,
  APILabelCounterT,
  APIManagedSilenceT,
} from "Models/APITypes";

const EmptyAPIResponse = (): APIAlertsResponseT => ({
  status: "success",
  error: "",
  timestamp: new Date().toISOString(),
  version: "fakeVersion",
  upstreams: {
    counters: { total: 1, healthy: 1, failed: 0 },
    clusters: { default: ["default"] },
    instances: [
      {
        name: "default",
        cluster: "default",
        clusterMembers: ["default"],
        uri: "http://localhost",
        publicURI: "http://localhost",
        error: "",
        version: "0.21.0",
        readonly: false,
        corsCredentials: "include",
        headers: {},
      },
    ],
  },
  silences: { default: {} },
  grids: [],
  receivers: ["by-cluster-service", "by-name"],
  totalAlerts: 0,
  colors: {},
  filters: [
    {
      text: "label=value",
      name: "label",
      matcher: "=",
      value: "value",
      hits: 0,
      isValid: true,
    },
  ],
  counters: [] as APILabelCounterT[],
  settings: {
    sorting: {
      grid: {
        order: "startsAt",
        reverse: false,
        label: "alertname",
      },
      valueMapping: {
        cluster: {
          dev: 3,
          prod: 1,
          staging: 2,
        },
        severity: {
          critical: 1,
          info: 3,
          warning: 2,
        },
      },
    },
    silenceForm: {
      strip: {
        labels: [],
      },
    },
    alertAcknowledgement: {
      enabled: false,
      durationSeconds: 900,
      author: "karma / author missing",
      comment: "ACK! Mock comment",
    },
    staticColorLabels: ["job"],
    annotationsDefaultHidden: false,
    annotationsHidden: [],
    annotationsVisible: [],
    annotationsEnableHTML: false,
    historyEnabled: true,
  },
  authentication: {
    username: "",
    enabled: false,
  },
});

const MockAPIResponse = (): APIAlertsResponseT => {
  const response = EmptyAPIResponse();
  response.grids = [
    {
      labelName: "",
      labelValue: "",
      alertGroups: [
        MockAlertGroup(
          { alertname: "foo" },
          [MockAlert([], { instance: "foo" }, "suppressed")],
          [],
          { cluster: "dev" },
          {}
        ),
      ],
      stateCount: {
        unprocessed: 1,
        suppressed: 2,
        active: 3,
      },
    },
  ];
  response.totalAlerts = 25;
  return response;
};

const MockSilenceResponse = (
  cluster: string,
  count: number
): APIManagedSilenceT[] => {
  const silences: APIManagedSilenceT[] = [];
  for (let index = 1; index <= count; index++) {
    const silence = MockSilence();
    silence.id = `silence${index}`;
    silences.push({
      cluster: cluster,
      alertCount: 123,
      isExpired: index < 3,
      silence: silence,
    });
  }
  return silences;
};

export { EmptyAPIResponse, MockAPIResponse, MockSilenceResponse };
