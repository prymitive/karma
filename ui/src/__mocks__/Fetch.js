import { MockAlert, MockAlertGroup, MockSilence } from "./Alerts";

const EmptyAPIResponse = () => ({
  status: "success",
  timestamp: new Date().toISOString(),
  version: "fakeVersion",
  upstreams: {
    counters: { total: 1, healthy: 1, failed: 0 },
    instances: [{ name: "default", uri: "file:///mock", error: "" }],
  },
  silences: { default: {} },
  grids: [],
  receivers: ["by-cluster-service", "by-name"],
  totalAlerts: 0,
  colors: {},
  filters: [
    {
      text: "label=value",
      name: "label",
      matcher: "=",
      value: "value",
      hits: 0,
      isValid: true,
    },
  ],
  counters: {},
  settings: {
    sorting: {
      grid: {
        order: "startsAt",
        reverse: false,
        label: "alertname",
      },
      valueMapping: {
        cluster: {
          dev: 3,
          prod: 1,
          staging: 2,
        },
        severity: {
          critical: 1,
          info: 3,
          warning: 2,
        },
      },
    },
    silenceForm: {
      strip: {
        labels: [],
      },
    },
    alertAcknowledgement: {
      enabled: false,
      durationSeconds: 900,
      author: "karma / author missing",
      commentPrefix: "",
    },
    staticColorLabels: ["job"],
    annotationsDefaultHidden: false,
    annotationsHidden: [],
    annotationsVisible: [],
  },
});

const MockAPIResponse = () => {
  const response = EmptyAPIResponse();
  response.grids = [
    {
      labelName: "",
      labelValue: "",
      alertGroups: [
        MockAlertGroup(
          { alertname: "foo" },
          [MockAlert([], { instance: "foo" }, "suppressed")],
          [],
          { cluster: "dev" },
          {}
        ),
      ],
      stateCount: {
        unprocessed: 1,
        suppressed: 2,
        active: 3,
      },
    },
  ];
  response.totalAlerts = 25;
  return response;
};

const MockSilenceResponse = (cluster, count) => {
  let silences = [];
  for (var index = 1; index <= count; index++) {
    const silence = MockSilence();
    silence.id = `silence${index}`;
    silences.push({
      cluster: cluster,
      alertCount: 123,
      isExpired: index < 3,
      silence: silence,
    });
  }
  return silences;
};

export { EmptyAPIResponse, MockAPIResponse, MockSilenceResponse };
