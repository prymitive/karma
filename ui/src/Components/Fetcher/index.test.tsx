import { act } from "react-dom/test-utils";

import { mount } from "enzyme";

import fetchMock from "fetch-mock";

import toDiffableHtml from "diffable-html";

import { advanceTo, advanceBy, clear } from "jest-date-mock";

import { EmptyAPIResponse } from "__fixtures__/Fetch";

import { AlertStore } from "Stores/AlertStore";
import { Settings } from "Stores/Settings";

import { Fetcher, Dots } from ".";

let alertStore: AlertStore;
let settingsStore: Settings;
let fetchSpy: any;
let requestAnimationFrameSpy: any;

beforeEach(() => {
  jest.useFakeTimers();
  advanceTo(new Date(Date.UTC(2000, 1, 1, 0, 0, 0)));

  alertStore = new AlertStore(["label=value"]);
  fetchSpy = jest
    .spyOn(alertStore, "fetchWithThrottle")
    .mockImplementation(() => {
      alertStore.status.setIdle();
      return new Promise((success) => {
        success();
      });
    });

  settingsStore = new Settings(null);
  settingsStore.fetchConfig.setInterval(30);

  requestAnimationFrameSpy = jest
    .spyOn(window, "requestAnimationFrame")
    .mockImplementation((cb: any) => {
      cb();
      return 0;
    });
});

afterEach(() => {
  alertStore.status.resume();
  requestAnimationFrameSpy.mockRestore();
  jest.clearAllTimers();
  jest.clearAllMocks();
  jest.restoreAllMocks();
  clear();
  fetchMock.reset();
});

const MockEmptyAPIResponseWithoutFilters = () => {
  const response = EmptyAPIResponse();
  response.filters = [];
  fetchMock.reset();
  fetchMock.mock("*", {
    status: 200,
    body: JSON.stringify(response),
  });
};

describe("<Fetcher />", () => {
  it("changing interval changes how often fetch is called", () => {
    settingsStore.fetchConfig.setInterval(1);
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledTimes(1);

    advanceBy(3 * 1000);
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(fetchSpy).toHaveBeenCalledTimes(2);

    settingsStore.fetchConfig.setInterval(600);

    advanceBy(3 * 1000);
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(fetchSpy).toHaveBeenCalledTimes(2);

    advanceBy(32 * 1000);
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(fetchSpy).toHaveBeenCalledTimes(2);

    advanceBy(62 * 1000);
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(fetchSpy).toHaveBeenCalledTimes(2);

    advanceBy(602 * 1000);
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(fetchSpy).toHaveBeenCalledTimes(3);
  });

  it("calls alertStore.fetchWithThrottle on mount", () => {
    const fetchSpy = jest.spyOn(alertStore, "fetchWithThrottle");
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledTimes(1);
  });

  it("calls alertStore.fetchWithThrottle again after filter change", () => {
    MockEmptyAPIResponseWithoutFilters();
    const fetchSpy = jest.spyOn(alertStore, "fetchWithThrottle");
    const tree = mount(
      <Fetcher alertStore={alertStore} settingsStore={settingsStore} />
    );
    alertStore.filters.setFilterValues([]);
    tree.setProps({});
    expect(fetchSpy).toHaveBeenCalledTimes(2);
  });

  it("keeps calling alertStore.fetchWithThrottle every minute", () => {
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledTimes(1);

    advanceBy(62 * 1000);
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(fetchSpy).toHaveBeenCalledTimes(2);

    advanceBy(62 * 1000);
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(fetchSpy).toHaveBeenCalledTimes(3);

    advanceBy(62 * 1000);
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(fetchSpy).toHaveBeenCalledTimes(4);
  });

  it("calls alertStore.fetchWithThrottle with empty sort arguments when sortOrder=default", () => {
    MockEmptyAPIResponseWithoutFilters();
    const fetchSpy = jest.spyOn(alertStore, "fetchWithThrottle");
    settingsStore.gridConfig.setSortOrder("default");
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledWith("", false, "", "", "");
  });

  it("calls alertStore.fetchWithThrottle with correct sort arguments when sortOrder=disabled reverseSort=false", () => {
    MockEmptyAPIResponseWithoutFilters();
    const fetchSpy = jest.spyOn(alertStore, "fetchWithThrottle");
    settingsStore.gridConfig.setSortOrder("disabled");
    settingsStore.gridConfig.setSortReverse(false);
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledWith("", false, "disabled", "", "");
  });

  it("calls alertStore.fetchWithThrottle with correct sort arguments when sortOrder=disabled reverseSort=true", () => {
    MockEmptyAPIResponseWithoutFilters();
    const fetchSpy = jest.spyOn(alertStore, "fetchWithThrottle");
    settingsStore.gridConfig.setSortOrder("disabled");
    settingsStore.gridConfig.setSortReverse(true);
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledWith("", false, "disabled", "", "");
  });

  it("calls alertStore.fetchWithThrottle with correct sort arguments when sortOrder=startsAt reverseSort=false", () => {
    MockEmptyAPIResponseWithoutFilters();
    const fetchSpy = jest.spyOn(alertStore, "fetchWithThrottle");
    settingsStore.gridConfig.setSortOrder("startsAt");
    settingsStore.gridConfig.setSortReverse(false);
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledWith("", false, "startsAt", "", "0");
  });

  it("calls alertStore.fetchWithThrottle with correct sort arguments when sortOrder=startsAt reverseSort=true", () => {
    MockEmptyAPIResponseWithoutFilters();
    const fetchSpy = jest.spyOn(alertStore, "fetchWithThrottle");
    settingsStore.gridConfig.setSortOrder("startsAt");
    settingsStore.gridConfig.setSortReverse(true);
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledWith("", false, "startsAt", "", "1");
  });

  it("calls alertStore.fetchWithThrottle with correct sort arguments when sortOrder=label sortLabel=cluster reverseSort=false", () => {
    MockEmptyAPIResponseWithoutFilters();
    const fetchSpy = jest.spyOn(alertStore, "fetchWithThrottle");
    settingsStore.gridConfig.setSortOrder("label");
    settingsStore.gridConfig.setSortLabel("cluster");
    settingsStore.gridConfig.setSortReverse(false);
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledWith("", false, "label", "cluster", "0");
  });

  it("calls alertStore.fetchWithThrottle with correct sort arguments when sortOrder=label sortLabel=job reverseSort=true", () => {
    MockEmptyAPIResponseWithoutFilters();
    const fetchSpy = jest.spyOn(alertStore, "fetchWithThrottle");
    settingsStore.gridConfig.setSortOrder("label");
    settingsStore.gridConfig.setSortLabel("job");
    settingsStore.gridConfig.setSortReverse(true);
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledWith("", false, "label", "job", "1");
  });

  it("calls alertStore.fetchWithThrottle with correct sort arguments when sortOrder=label sortLabel=instance reverseSort=null", () => {
    MockEmptyAPIResponseWithoutFilters();
    const fetchSpy = jest.spyOn(alertStore, "fetchWithThrottle");
    settingsStore.gridConfig.setSortOrder("label");
    settingsStore.gridConfig.setSortLabel("instance");
    settingsStore.gridConfig.setSortReverse(null);
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledWith("", false, "label", "instance", "");
  });

  it("calls alertStore.fetchWithThrottle with gridLabel=cluster gridSortReverse=false", () => {
    MockEmptyAPIResponseWithoutFilters();
    const fetchSpy = jest.spyOn(alertStore, "fetchWithThrottle");
    settingsStore.gridConfig.setSortOrder("default");
    settingsStore.multiGridConfig.setGridLabel("cluster");
    settingsStore.multiGridConfig.setGridSortReverse(false);
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledWith("cluster", false, "", "", "");
  });

  it("calls alertStore.fetchWithThrottle with gridLabel=cluster gridSortReverse=true", () => {
    MockEmptyAPIResponseWithoutFilters();
    const fetchSpy = jest.spyOn(alertStore, "fetchWithThrottle");
    settingsStore.gridConfig.setSortOrder("default");
    settingsStore.multiGridConfig.setGridLabel("cluster");
    settingsStore.multiGridConfig.setGridSortReverse(true);
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledWith("cluster", true, "", "", "");
  });

  it("calls alertStore.fetchWithThrottle with gridLabel= gridSortReverse=true", () => {
    MockEmptyAPIResponseWithoutFilters();
    const fetchSpy = jest.spyOn(alertStore, "fetchWithThrottle");
    settingsStore.gridConfig.setSortOrder("default");
    settingsStore.multiGridConfig.setGridLabel("");
    settingsStore.multiGridConfig.setGridSortReverse(true);
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledWith("", true, "", "", "");
  });

  it("internal timer is null after unmount", () => {
    const tree = mount(
      <Fetcher alertStore={alertStore} settingsStore={settingsStore} />
    );
    expect(fetchSpy).toHaveBeenCalledTimes(1);

    act(() => {
      tree.unmount();
    });
    expect(fetchSpy).toHaveBeenCalledTimes(1);

    act(() => {
      settingsStore.gridConfig.setSortReverse(
        !settingsStore.gridConfig.config.reverseSort
      );
    });
    expect(fetchSpy).toHaveBeenCalledTimes(1);
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(fetchSpy).toHaveBeenCalledTimes(1);
  });

  it("doesn't fetch on mount when paused", () => {
    alertStore.status.pause();
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    expect(fetchSpy).toHaveBeenCalledTimes(0);
  });

  it("doesn't fetch on update when paused", () => {
    alertStore.status.pause();
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    settingsStore.gridConfig.setSortReverse(
      !settingsStore.gridConfig.config.reverseSort
    );
    expect(fetchSpy).toHaveBeenCalledTimes(0);
  });

  it("fetches on update when resumed", () => {
    alertStore.status.pause();
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    alertStore.status.resume();
    settingsStore.gridConfig.setSortReverse(
      !settingsStore.gridConfig.config.reverseSort
    );
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(fetchSpy).toHaveBeenCalledTimes(1);
  });

  it("fetches on resume", () => {
    alertStore.status.pause();
    mount(<Fetcher alertStore={alertStore} settingsStore={settingsStore} />);
    alertStore.status.resume();
    advanceBy(2 * 1000);
    act(() => {
      jest.runOnlyPendingTimers();
    });
    expect(fetchSpy).toHaveBeenCalledTimes(1);
  });
});

describe("<Fetcher /> children", () => {
  it("renders Dots when countdown is in progress", () => {
    const tree = mount(
      <Fetcher alertStore={alertStore} settingsStore={settingsStore} />
    );
    expect(tree.find("div.components-fetcher")).toHaveLength(1);
  });

  it("doesn't render any children when upgrade is needed", () => {
    act(() => {
      alertStore.info.setUpgradeNeeded(true);
    });
    const tree = mount(
      <Fetcher alertStore={alertStore} settingsStore={settingsStore} />
    );
    expect(tree.find("div.navbar-brand").children()).toHaveLength(0);
  });

  it("renders PauseButton when paused", () => {
    const tree = mount(
      <Fetcher alertStore={alertStore} settingsStore={settingsStore} />
    );
    act(() => {
      alertStore.status.pause();
    });
    expect(toDiffableHtml(tree.html())).toMatch(/fa-pause/);
  });

  it("renders PauseButton when paused and hovered", () => {
    const tree = mount(
      <Fetcher alertStore={alertStore} settingsStore={settingsStore} />
    );
    act(() => {
      alertStore.status.pause();
    });
    tree.find(".navbar-brand").simulate("mouseenter");
    tree.update();
    expect(toDiffableHtml(tree.html())).toMatch(/fa-pause/);

    tree.find(".navbar-brand").simulate("mouseleave");
    tree.update();
    expect(toDiffableHtml(tree.html())).toMatch(/fa-pause/);
  });

  it("renders PlayButton when hovered", () => {
    const tree = mount(
      <Fetcher alertStore={alertStore} settingsStore={settingsStore} />
    );
    tree.find(".navbar-brand").simulate("mouseenter");
    tree.update();
    expect(toDiffableHtml(tree.html())).toMatch(/fa-play/);

    tree.find(".navbar-brand").simulate("mouseleave");
    tree.update();
    expect(tree.find("div.components-fetcher")).toHaveLength(1);
  });
});

describe("<Dots />", () => {
  it("matches snapshot", () => {
    const tree = mount(<Dots alertStore={alertStore} dots={8} />);
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("adds 'fetching' class when fetching data", () => {
    act(() => {
      alertStore.status.setFetching();
    });
    const tree = mount(<Dots alertStore={alertStore} dots={8} />);
    expect(tree.find("div.components-fetcher").hasClass("fetching")).toBe(true);
  });

  it("adds 'processing' class when processing fetched data", () => {
    act(() => {
      alertStore.status.setProcessing();
    });
    const tree = mount(<Dots alertStore={alertStore} dots={8} />);
    expect(tree.find("div.components-fetcher").hasClass("processing")).toBe(
      true
    );
  });

  it("adds 'retrying' class when fetch needs a retry", () => {
    act(() => {
      alertStore.info.setIsRetrying();
    });
    const tree = mount(<Dots alertStore={alertStore} dots={8} />);
    expect(tree.find("div.components-fetcher").hasClass("retrying")).toBe(true);
  });
});
