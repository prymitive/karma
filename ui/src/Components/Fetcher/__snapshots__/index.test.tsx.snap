// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dots /> matches snapshot 1`] = `
"
<div class=\\"cursor-pointer components-fetcher   \\">
  <div class=\\"dot  visible\\">
  </div>
  <div class=\\"dot  visible\\">
  </div>
  <div class=\\"dot  visible\\">
  </div>
  <div class=\\"dot  visible\\">
  </div>
  <div class=\\"dot dot-middle visible\\">
  </div>
  <div class=\\"dot  visible\\">
  </div>
  <div class=\\"dot  visible\\">
  </div>
  <div class=\\"dot  visible\\">
  </div>
  <div class=\\"dot  hidden\\">
  </div>
</div>
"
`;
